
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="CBICall: CNAG Biomedical Informatics framework for variant Calling">
      
      
        <meta name="author" content="Manuel Rueda, PhD">
      
      
      
        <link rel="prev" href="../quickstart/">
      
      
        <link rel="next" href="../../help/faq/">
      
      
      <link rel="icon" href="../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.22">
    
    
      
        <title>Usage - CBICall Documentation</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.84d31ad4.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../css/theme_overrides.css">
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="blue-grey" data-md-color-accent="amber">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#essential-parameters" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="CBICall Documentation" class="md-header__button md-logo" aria-label="CBICall Documentation" data-md-component="logo">
      
  <img src="../../img/cbicall-logo.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            CBICall Documentation
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Usage
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="default" data-md-color-primary="blue-grey" data-md-color-accent="amber"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 2a7 7 0 0 0-7 7c0 2.38 1.19 4.47 3 5.74V17a1 1 0 0 0 1 1h6a1 1 0 0 0 1-1v-2.26c1.81-1.27 3-3.36 3-5.74a7 7 0 0 0-7-7M9 21a1 1 0 0 0 1 1h4a1 1 0 0 0 1-1v-1H9z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="slate" data-md-color-primary="blue-grey" data-md-color-accent="amber"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 2a7 7 0 0 1 7 7c0 2.38-1.19 4.47-3 5.74V17a1 1 0 0 1-1 1H9a1 1 0 0 1-1-1v-2.26C6.19 13.47 5 11.38 5 9a7 7 0 0 1 7-7M9 21v-1h6v1a1 1 0 0 1-1 1h-4a1 1 0 0 1-1-1m3-17a5 5 0 0 0-5 5c0 2.05 1.23 3.81 3 4.58V16h4v-2.42c1.77-.77 3-2.53 3-4.58a5 5 0 0 0-5-5"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
          <a href="javascript:void(0)" class="md-search__icon md-icon" title="Share" aria-label="Share" data-clipboard data-clipboard-text="" data-md-component="search-share" tabindex="-1">
            
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7s-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9a3 3 0 0 0-3 3 3 3 0 0 0 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.15c-.05.21-.08.43-.08.66 0 1.61 1.31 2.91 2.92 2.91s2.92-1.3 2.92-2.91A2.92 2.92 0 0 0 18 16.08"/></svg>
          </a>
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/CNAG-Biomedical-Informatics/cbicall" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M202.1 328.7c0 20.9-10.9 55.1-36.7 55.1s-36.7-34.2-36.7-55.1 10.9-55.1 36.7-55.1 36.7 34.2 36.7 55.1M496 278.2c0 31.9-3.2 65.7-17.5 95-37.9 76.6-142.1 74.8-216.7 74.8-75.8 0-186.2 2.7-225.6-74.8-14.6-29-20.2-63.1-20.2-95 0-41.9 13.9-81.5 41.5-113.6-5.2-15.8-7.7-32.4-7.7-48.8 0-21.5 4.9-32.3 14.6-51.8 45.3 0 74.3 9 108.8 36 29-6.9 58.8-10 88.7-10 27 0 54.2 2.9 80.4 9.2 34-26.7 63-35.2 107.8-35.2 9.8 19.5 14.6 30.3 14.6 51.8 0 16.4-2.6 32.7-7.7 48.2 27.5 32.4 39 72.3 39 114.2m-64.3 50.5c0-43.9-26.7-82.6-73.5-82.6-18.9 0-37 3.4-56 6-14.9 2.3-29.8 3.2-45.1 3.2-15.2 0-30.1-.9-45.1-3.2-18.7-2.6-37-6-56-6-46.8 0-73.5 38.7-73.5 82.6 0 87.8 80.4 101.3 150.4 101.3h48.2c70.3 0 150.6-13.4 150.6-101.3m-82.6-55.1c-25.8 0-36.7 34.2-36.7 55.1s10.9 55.1 36.7 55.1 36.7-34.2 36.7-55.1-10.9-55.1-36.7-55.1"/></svg>
  </div>
  <div class="md-source__repository">
    CBICall
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



  

<nav class="md-nav md-nav--primary md-nav--integrated" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="CBICall Documentation" class="md-nav__button md-logo" aria-label="CBICall Documentation" data-md-component="logo">
      
  <img src="../../img/cbicall-logo.png" alt="logo">

    </a>
    CBICall Documentation
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/CNAG-Biomedical-Informatics/cbicall" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M202.1 328.7c0 20.9-10.9 55.1-36.7 55.1s-36.7-34.2-36.7-55.1 10.9-55.1 36.7-55.1 36.7 34.2 36.7 55.1M496 278.2c0 31.9-3.2 65.7-17.5 95-37.9 76.6-142.1 74.8-216.7 74.8-75.8 0-186.2 2.7-225.6-74.8-14.6-29-20.2-63.1-20.2-95 0-41.9 13.9-81.5 41.5-113.6-5.2-15.8-7.7-32.4-7.7-48.8 0-21.5 4.9-32.3 14.6-51.8 45.3 0 74.3 9 108.8 36 29-6.9 58.8-10 88.7-10 27 0 54.2 2.9 80.4 9.2 34-26.7 63-35.2 107.8-35.2 9.8 19.5 14.6 30.3 14.6 51.8 0 16.4-2.6 32.7-7.7 48.2 27.5 32.4 39 72.3 39 114.2m-64.3 50.5c0-43.9-26.7-82.6-73.5-82.6-18.9 0-37 3.4-56 6-14.9 2.3-29.8 3.2-45.1 3.2-15.2 0-30.1-.9-45.1-3.2-18.7-2.6-37-6-56-6-46.8 0-73.5 38.7-73.5 82.6 0 87.8 80.4 101.3 150.4 101.3h48.2c70.3 0 150.6-13.4 150.6-101.3m-82.6-55.1c-25.8 0-36.7 34.2-36.7 55.1s10.9 55.1 36.7 55.1 36.7-34.2 36.7-55.1-10.9-55.1-36.7-55.1"/></svg>
  </div>
  <div class="md-source__repository">
    CBICall
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    🏠 Introduction
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
      
        
        
      
    
    
    <li class="md-nav__item md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" >
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="">
            
  
  
  <span class="md-ellipsis">
    ⬇️ Download & Installation
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            ⬇️ Download & Installation
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../download-and-installation/non-containerized/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    🛠 Non-containerized
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../download-and-installation/docker-based/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    🐳 Docker-based
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
      
        
        
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" checked>
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="">
            
  
  
  <span class="md-ellipsis">
    🚀 Getting Started
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            🚀 Getting Started
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../quickstart/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Quickstart
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    Usage
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    Usage
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#essential-parameters" class="md-nav__link">
    <span class="md-ellipsis">
      Essential Parameters
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#optional-parameters-currently-unused" class="md-nav__link">
    <span class="md-ellipsis">
      Optional Parameters (Currently Unused)
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#example-commands" class="md-nav__link">
    <span class="md-ellipsis">
      Example Commands
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#addendum-nomenclature-guidelines" class="md-nav__link">
    <span class="md-ellipsis">
      ADDENDUM: Nomenclature Guidelines
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#directory-naming" class="md-nav__link">
    <span class="md-ellipsis">
      Directory Naming
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#subdirectory-naming" class="md-nav__link">
    <span class="md-ellipsis">
      Subdirectory Naming
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#fastq-naming-convention" class="md-nav__link">
    <span class="md-ellipsis">
      FASTQ Naming Convention
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#capture-kits" class="md-nav__link">
    <span class="md-ellipsis">
      Capture Kits
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
      
        
        
      
    
    
    <li class="md-nav__item md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" >
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="">
            
  
  
  <span class="md-ellipsis">
    ❓ Help
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            ❓ Help
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../help/faq/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    🤔 FAQs
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
      
        
        
      
    
    
    <li class="md-nav__item md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" >
        
          
          <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="">
            
  
  
  <span class="md-ellipsis">
    ℹ️ About
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            ℹ️ About
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../about/about/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    About
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../about/citation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Citation
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  
  


<div align="center">
  <a href="https://github.com/mrueda/cbicall">
    <img src="https://raw.githubusercontent.com/mrueda/cbicall/main/docs/img/cbicall-logo.png"
         width="350" alt="CBICall">
  </a>
  <p><em>CNAG Biomedical Informatics framework for variant Calling</em></p>
</div>

<p><a href="https://github.com/cnag-biomedical-informatics/cbicall/actions/workflows/build.yml"><img alt="Build" src="https://github.com/cnag-biomedical-informatics/cbicall/actions/workflows/build.yml/badge.svg" /></a>
<img alt="version" src="https://img.shields.io/badge/version-0.0.1-28a745" />
<a href="https://github.com/cnag-biomedical-informatics/cbicall/actions/workflows/docker-build-multi-arch.yml"><img alt="Docker Build" src="https://github.com/cnag-biomedical-informatics/cbicall/actions/workflows/docker-build-multi-arch.yml/badge.svg" /></a>
<a href="https://hub.docker.com/r/manuelrueda/cbicall/"><img alt="Docker Pulls" src="https://badgen.net/docker/pulls/manuelrueda/cbicall?icon=docker&amp;label=pulls" /></a>
<a href="https://hub.docker.com/r/manuelrueda/cbicall/"><img alt="Docker Image Size" src="https://badgen.net/docker/size/manuelrueda/cbicall?icon=docker&amp;label=image%20size" /></a>
<a href="https://github.com/cnag-biomedical-informatics/cbicall/actions/workflows/documentation.yml"><img alt="Documentation Status" src="https://github.com/cnag-biomedical-informatics/cbicall/actions/workflows/documentation.yml/badge.svg" /></a>
<a href="https://www.gnu.org/licenses/gpl-3.0"><img alt="License: GPL v3" src="https://img.shields.io/badge/License-GPL%20v3-blue.svg" /></a></p>
<hr />
<p><strong>📘 Documentation:</strong> <a href="https://cnag-biomedical-informatics.github.io/cbicall" target="_blank"><a href="https://cnag-biomedical-informatics.github.io/cbicall">cnag-biomedical-informatics.github.io/cbicall</a></a></p>
<p><strong>🐳 Docker Hub Image:</strong> <a href="https://hub.docker.com/r/manuelrueda/cbicall/tags" target="_blank"><a href="https://hub.docker.com/r/manuelrueda/cbicall/tags">hub.docker.com/r/manuelrueda/cbicall/tags</a></a></p>
<h1 id="table-of-contents">Table of contents<a class="headerlink" href="#table-of-contents" title="Permanent link">&para;</a></h1>
<ul>
<li>Description</li>
<li><a href="#name">Name</a></li>
<li><a href="#synopsis">Synopsis</a></li>
<li><a href="#summary">Summary</a></li>
<li><a href="#installation">Installation</a></li>
<li><a href="non-containerized/README.md">Non-Containerized</a></li>
<li><a href="docker/README.md">Containerized</a></li>
<li><a href="#how-to-run-cbicall">How to run cbicall</a></li>
<li><a href="#citation">Citation</a></li>
<li><a href="#author">Author</a></li>
<li><a href="#copyright-and-license">License</a></li>
</ul>
<h1 id="name">NAME<a class="headerlink" href="#name" title="Permanent link">&para;</a></h1>
<p>CBICall: CNAG Biomedical Informatics Framework for Variant Calling on Illumina DNA-seq (germline) NGS Data.</p>
<h1 id="synopsis">SYNOPSIS<a class="headerlink" href="#synopsis" title="Permanent link">&para;</a></h1>
<div class="language-text highlight"><pre><span></span><code>cbicall -p &lt;parameters_file.yaml&gt; -t &lt;n_threads&gt; [options]

Arguments:
  -p|param          Parameters input file (YAML)
  -t|threads        Number of CPUs/Cores/Threads

Options:
  -debug            Debugging level (from 1 to 5; 5 is maximum verbosity)
  -h|help           Brief help message
  -man              Full documentation
  -v                Show version information
  -verbose          Enable verbose output
  -nc|no-color      Do not print colors to STDOUT
  -ne|no-emoji      Do not print emojis to STDOUT
</code></pre></div>
<h1 id="summary">SUMMARY<a class="headerlink" href="#summary" title="Permanent link">&para;</a></h1>
<p>CBICall (<strong>C</strong>NAG <strong>B</strong>iomedical <strong>I</strong>nformatics framework for variant <strong>Call</strong>ing) is a computational framework designed for variant calling analysis using Illumina Next-Generation Sequencing (NGS) data.</p>
<h1 id="how-to-run-cbicall">HOW TO RUN CBICALL<a class="headerlink" href="#how-to-run-cbicall" title="Permanent link">&para;</a></h1>
<p>CBICall execution requires:</p>
<ul>
<li>
<p>Input Files</p>
<p>A folder containing Paired-End FASTQ files (e.g., <code>MA00001_exome/MA0000101P_ex/*{R1,R2}*fastq.gz</code>).</p>
<p>You have a <code>examples/input/</code> directory with input data that you can use for testing.</p>
</li>
<li>
<p>Parameters File</p>
<p>A YAML-formatted parameters file controlling pipeline execution.</p>
</li>
</ul>
<p>Below are the parameters that can be customized, along with their default values. Parameters must be separated from their values by whitespace or tabs.</p>
<h2 id="essential-parameters">Essential Parameters<a class="headerlink" href="#essential-parameters" title="Permanent link">&para;</a></h2>
<div class="language-text highlight"><pre><span></span><code>mode:            single  
pipeline:        wes          
sample:          undef        
sample_map:      undef
workflow_engine:   bash
gatk_version:      gatk3.5
cleanup_bam:       false
</code></pre></div>
<h2 id="optional-parameters-currently-unused">Optional Parameters (Currently Unused)<a class="headerlink" href="#optional-parameters-currently-unused" title="Permanent link">&para;</a></h2>
<div class="language-text highlight"><pre><span></span><code>organism:        Homo Sapiens        
technology:      Illumina HiSeq
</code></pre></div>
<p>CBICall will create a dedicated project directory (<code>cbicall_*</code>) to store analysis outputs. This design allows multiple independent runs concurrently without modifying original input files.</p>
<p>Below is a detailed description of key parameters:</p>
<ul>
<li>
<p><strong>cleanup_bam</strong></p>
<p>Set it to <code>true</code> to delete <code>01_bam/*.{bam,bai}</code>.</p>
</li>
<li>
<p><strong>gatk_version</strong></p>
<p>Supported values: <code>gatk3.5</code> or <code>gatk4.6</code>.</p>
</li>
<li>
<p><strong>mode</strong></p>
<p>Two modes are supported: <code>single</code> (default, for individual samples) and <code>cohort</code> (for family/cohort-based analyses).</p>
</li>
<li>
<p><strong>pipeline</strong></p>
<p>Specifies the analysis pipeline. Currently available options: <code>wes</code> (whole-exome sequencing) and <code>mit</code> (mitochondrial DNA analysis). Note: to run <code>cohort</code> analysis, first complete a <code>single</code> analysis for each sample.</p>
</li>
<li>
<p><strong>projectdir</strong></p>
<p>The prefix for dir name (e.g., 'cancer_sample_001'). Note that it can also contain a path (e.g., foo/cancer_sample_001).</p>
<p><strong>Note:</strong> Such directory will be always created below the <strong>sample</strong> directory. The script will automatically add an unique identifier to each job.</p>
</li>
<li>
<p><strong>sample</strong></p>
<p>Path (relative or absolute) to the directory containing FASTQ files for analysis. See the <code>examples</code> directory for detailed guidance.</p>
<p>Example:</p>
<p>examples/input/CNAG999_exome/CNAG99901P_ex</p>
</li>
<li>
<p><strong>sample_map</strong> (cohort-mode only)</p>
<p>Path (relative or absolute) to the file containing the sample ids and teh paths for the GVCF files</p>
<p>See example <a href="https://github.com/CNAG-Biomedical-Informatics/cbicall/tree/main/examples/input/sample_map.tsv">here</a></p>
</li>
<li>
<p><strong>workflow_engine</strong></p>
<p>Supported workflow engines: <code>bash</code> or <code>snakemake</code>.</p>
</li>
</ul>
<h2 id="example-commands">Example Commands<a class="headerlink" href="#example-commands" title="Permanent link">&para;</a></h2>
<div class="language-text highlight"><pre><span></span><code>$ bin/cbicall -p param_file.yaml -t 8
$ bin/cbicall -p param_file.yaml -t 4 -verbose
$ bin/cbicall -p param_file.yaml -t 16 &gt; log 2&gt;&amp;1
$ $path_to_cbicall/bin/cbicall -p param_file.yaml -t 8 -debug 5
</code></pre></div>
<p>Note: For Trio analyses, unique (de novo) variant rates for probands typically should be ~1%, and ~10% for parents. Significant deviations may indicate issues.</p>
<h2 id="addendum-nomenclature-guidelines">ADDENDUM: Nomenclature Guidelines<a class="headerlink" href="#addendum-nomenclature-guidelines" title="Permanent link">&para;</a></h2>
<p>All parts must follow a strict character count, and everything after the underscore is mandatory.</p>
<h2 id="directory-naming">Directory Naming<a class="headerlink" href="#directory-naming" title="Permanent link">&para;</a></h2>
<ul>
<li>
<p>Format: <code>[ProjectCode]_[SampleType]</code></p>
<ul>
<li>
<ul>
<li><code>ProjectCode</code>: Exactly 7 characters [a-zA-Z0-9] (e.g., <code>MA99999</code>)</li>
</ul>
</li>
<li>
<ul>
<li><code>SampleType</code>: Must be <code>exome</code> (5 characters)</li>
</ul>
</li>
</ul>
<p>Example:</p>
<div class="language-text highlight"><pre><span></span><code>MA99999_exome
</code></pre></div>
<p>Total: 7 + 1 + 5 = 13 characters.</p>
</li>
</ul>
<h2 id="subdirectory-naming">Subdirectory Naming<a class="headerlink" href="#subdirectory-naming" title="Permanent link">&para;</a></h2>
<ul>
<li>
<p>Format: <code>[ProjectCode][SampleID][Role]_[SampleTypeShort]</code></p>
<ul>
<li>
<ul>
<li><code>ProjectCode</code>: 7 characters ([a-zA-Z0-9] e.g., <code>MA99999</code>)</li>
</ul>
</li>
<li>
<ul>
<li><code>SampleID</code>: 2 characters (e.g., <code>01</code>)</li>
</ul>
</li>
<li>
<ul>
<li><code>Role</code>: 1 character (e.g., <code>P</code> for Proband, <code>F</code> for Father, <code>M</code> for Mother)</li>
</ul>
</li>
<li>
<ul>
<li><code>SampleTypeShort</code>: Must be <code>ex</code> (2 characters)</li>
</ul>
</li>
</ul>
<p>Example:</p>
<div class="language-text highlight"><pre><span></span><code>MA9999901F_ex
</code></pre></div>
<p>Total: 7 + 2 + 1 + 1 + 2 = 13 characters (excluding any file extension).</p>
</li>
</ul>
<h2 id="fastq-naming-convention">FASTQ Naming Convention<a class="headerlink" href="#fastq-naming-convention" title="Permanent link">&para;</a></h2>
<p>This convention is adapted from the following document:</p>
<p><a href="https://support.illumina.com/help/BaseSpace_Sequence_Hub_OLH_009008_2/Source/Informatics/BS/NamingConvention_FASTQ-files-swBS.htm">Document</a></p>
<p>We have added a custom suffix to indicate sequencing type:</p>
<div class="language-text highlight"><pre><span></span><code>- _ex for exome sequencing
- _wg for whole genome sequencing
</code></pre></div>
<p>Example:</p>
<div class="language-text highlight"><pre><span></span><code>MA0004701P_ex_S5_L001_R1_001.fastq.gz
</code></pre></div>
<p>In summary, you need to have something like this:</p>
<div class="language-text highlight"><pre><span></span><code>MA00001_exome/MA0000101P_ex/MA0000101P_ex_S1_L001_R?_001.fastq.gz
</code></pre></div>
<h1 id="system-requirements">SYSTEM REQUIREMENTS<a class="headerlink" href="#system-requirements" title="Permanent link">&para;</a></h1>
<p>CBICall is optimized for multi-core Linux desktop, workstation, or server environments. Snakemake-based workflows can also be adapted for HPC clusters.</p>
<p>Recommended specifications:</p>
<div class="language-text highlight"><pre><span></span><code>* Works in amd64 and arm64 archs (M-based Macs).
* Ideally a Debian-based distribution (Ubuntu or Mint), but any other (e.g., CentOS, OpenSUSE) should do as well (untested).
* &gt;= 8 GB RAM.
* &gt;= 4 CPU cores (Intel i7 or Xeon preferred).
* &gt;= 250 GB HDD space.
* Perl &gt;= 5.36 and required CPAN modules (install via C&lt;cpanm --notest --installdeps .&gt;).
* Java 8 (install via C&lt;sudo apt install openjdk-8-jdk&gt;).
* Snakemake (install via C&lt;pip3 install -r requirements.txt&gt;).
</code></pre></div>
<p>Perl scripts in CBICall use minimal RAM (~2% of a 16 GB system). Genome mapping with BWA benefits from higher memory but lacks built-in RAM limits. Its usage depends on thread count and reference size. To constrain BWA's memory, external tools like shell <code>ulimit</code> are required. In contrast, GATK and Picard default to 8 GB RAM, adjustable via the configuration file.</p>
<p>Parallel execution is supported but does not scale linearly. Optimal performance is achieved using ~ 4 threads per task. For example, with 12 cores, running 3 tasks in parallel with 4 cores each is typically more efficient than one task with all 12 cores. See example in figure below:</p>
<p><img alt="Time" src="https://github.com/CNAG-Biomedical-Informatics/cbicall/blob/main/docs/img/run-time.png" /></p>
<p>Unit/integration tests are conducted manually by verifying CSV and VCF outputs against established test datasets.</p>
<h1 id="supported-pipelines">SUPPORTED PIPELINES<a class="headerlink" href="#supported-pipelines" title="Permanent link">&para;</a></h1>
<p>The following table shows valid pipeline and mode combinations for each GATK version:</p>
<table>
<thead>
<tr>
<th>GATK Version</th>
<th>wes_single</th>
<th>wes_cohort</th>
<th>wgs_single</th>
<th>wgs_cohort</th>
<th>mit_single</th>
<th>mit_cohort</th>
</tr>
</thead>
<tbody>
<tr>
<td>gatk3.5</td>
<td>+</td>
<td>+</td>
<td>-</td>
<td>-</td>
<td>+</td>
<td>+</td>
</tr>
<tr>
<td>gatk4.6</td>
<td>+</td>
<td>+</td>
<td>+</td>
<td>+</td>
<td>-</td>
<td>-</td>
</tr>
</tbody>
</table>
<p>Date: Oct-2025</p>
<h2 id="capture-kits">Capture Kits<a class="headerlink" href="#capture-kits" title="Permanent link">&para;</a></h2>
<p>* For GATK version 3.5: Exome capture is based on Agilent SureSelect.</p>
<p>* For GATK version 4.6: Exome and WGS reference is based on the GATK bundle (b37).</p>
<h1 id="common-errors-and-troubleshooting">COMMON ERRORS AND TROUBLESHOOTING<a class="headerlink" href="#common-errors-and-troubleshooting" title="Permanent link">&para;</a></h1>
<ul>
<li>
<p>GATK|Picard Errors (wes_single.sh or wes_cohort.sh)</p>
<ul>
<li>
<p>Error: <code>NaN LOD value assigned</code> in recalibration steps.</p>
<p>Occurs due to insufficient INDEL variants (typically fewer than 8000) for negative model training. The default threshold is 8000.</p>
<p>Solution: Increase minimum INDEL count (e.g., to &gt;8000) in relevant pipeline scripts. Only rerun failed samples.</p>
</li>
<li>
<p>Error: <code>there aren't enough columns for line ... dbsnp_137.hg19.vcf</code></p>
<p>Solution: Remove the problematic line from the VCF file and document changes in a README file.</p>
</li>
<li>
<p>Error: <code>Error parsing text SAM file. Not enough fields; File /dev/stdin; Line 105120626...</code></p>
<p>Certain SRA-derived or dbGaP datasets can contain duplicate reads.</p>
<p>When piping BWA output into <code>AddOrReplaceReadGroups</code>, you may need to remove secondary (<code>0x100</code>) and supplementary (<code>0x800</code>) alignments.</p>
<p>This can prevent duplicate-read collisions in downstream Picard/GATK steps.</p>
<p><strong>Solution</strong>: Uncomment the following line in <code>wes_single.sh</code>:</p>
<p><code>| $SAM view -bSh -F 0x900 -</code>
- <strong>MTOOLBOX Errors</strong></p>
</li>
</ul>
<p>- Failure related to unsupported N_CIGAR:<br />
  Add flag <code>--filter_reads_with_N_cigar</code> in Mtoolbox.sh (line ~386).</p>
<p>- Samples with coverage below ~10x yield unreliable heteroplasmy fractions (HF). Extremely low coverage (&lt;10x) can render HFs meaningless, despite generally consistent results across widely varying coverage levels.</p>
</li>
</ul>
<h1 id="citation">CITATION<a class="headerlink" href="#citation" title="Permanent link">&para;</a></h1>
<p>To be determined.</p>
<h1 id="author">AUTHOR<a class="headerlink" href="#author" title="Permanent link">&para;</a></h1>
<p>Written by Manuel Rueda (mrueda). GitHub repository: <a href="https://github.com/CNAG-Biomedical-Informatics/cbicall">https://github.com/CNAG-Biomedical-Informatics/cbicall</a>.</p>
<h1 id="copyright-and-license">COPYRIGHT AND LICENSE<a class="headerlink" href="#copyright-and-license" title="Permanent link">&para;</a></h1>
<p>Please see the included LICENSE file for distribution and usage terms.</p>












                
              </article>
            </div>
          
          
  <script>var tabs=__md_get("__tabs");if(Array.isArray(tabs))e:for(var set of document.querySelectorAll(".tabbed-set")){var labels=set.querySelector(".tabbed-labels");for(var tab of tabs)for(var label of labels.getElementsByTagName("label"))if(label.innerText.trim()===tab){var input=document.getElementById(label.htmlFor);input.checked=!0;continue e}}</script>

<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      &copy; Copyright 2024-2025, Manuel Rueda, CNAG
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
        
<div class="md-social">
  
    
    
    
    
      
      
    
    <a href="https://github.com/CNAG-Biomedical-Informatics" target="_blank" rel="noopener" title="github.com" class="md-social__link">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M173.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M252.8 8C114.1 8 8 113.3 8 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C436.2 457.8 504 362.9 504 252 504 113.3 391.5 8 252.8 8M105.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg>
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      <script id="__config" type="application/json">{"base": "../..", "features": ["content.code.select", "content.code.copy", "content.tabs.link", "search.highlight", "search.share", "navigation.tracking", "navigation.sections", "toc.integrate"], "search": "../../assets/javascripts/workers/search.973d3a69.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../../assets/javascripts/bundle.f55a23d4.min.js"></script>
      
        <script src="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.7/katex.min.js"></script>
      
        <script src="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.7/contrib/auto-render.min.js"></script>
      
        <script src="../../js/vendor/katex.js"></script>
      
        <script src="../../js/theme_overrides.js"></script>
      
    
  </body>
</html>